pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,242 @@
+package delegatedidentity
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	delegatedidentityv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/agent/delegatedidentity/v1"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/spiffe/spire/pkg/agent/api/rpccontext"
+	workload_attestor "github.com/spiffe/spire/pkg/agent/attestor/workload"
+	"github.com/spiffe/spire/pkg/agent/endpoints"
+	"github.com/spiffe/spire/pkg/agent/manager"
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/proto/spire/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RegisterService registers the delegated identity service on the provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	delegatedidentityv1.RegisterDelegatedIdentityServer(s, service)
+}
+
+type attestor interface {
+	Attest(ctx context.Context) ([]*common.Selector, error)
+}
+
+type Config struct {
+	Log             logrus.FieldLogger
+	Manager         manager.Manager
+	Attestor        workload_attestor.Attestor
+	AuthorizedUsers []string
+}
+
+func New(config Config) *Service {
+	authorizedUsers := map[string]bool{}
+
+	for _, user := range config.AuthorizedUsers {
+		authorizedUsers[user] = true
+	}
+
+	return &Service{
+		manager:         config.Manager,
+		attestor:        endpoints.PeerTrackerAttestor{Attestor: config.Attestor},
+		authorizedUsers: authorizedUsers,
+	}
+}
+
+// Service implements the delegated identity server
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor attestor
+
+	// Spiffe IDs of users that are authorized to use this API
+	authorizedUsers map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized users map.
+func (s *Service) isCallerAuthorized(ctx context.Context, log logrus.FieldLogger) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return false, status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+	numRegisteredIDs := len(identities)
+
+	if numRegisteredIDs == 0 {
+		log.Error("no identity issued")
+		return false, status.Error(codes.PermissionDenied, "no identity issued")
+	}
+
+	for _, identity := range identities {
+		if _, ok := s.authorizedUsers[identity.Entry.SpiffeId]; ok {
+			return true, nil
+		}
+	}
+
+	// caller has identity associeted with but none is authorized
+	log.Error("Permission denied: caller not configured as an authorized delegate. Number of registers issued: ", numRegisteredIDs)
```

Comment:
We should use a structured field for this

Other log lines use `;` rather than `:`

```suggestion
	log.WithFields(log.Fields{
	    "num_registered_ids": numRegisteredIDs,
	    "default_id": identities[0].Entry.SpiffeId,
	}).Error("Permission denied; caller not configured as an authorized delegate.)
```

Reply:
@rscampos looks like this is still pending?